packages/client/hmi-client/src/components/responsive-matrix/matrix.vue

```diff
@@ -0,0 +1,902 @@
+		});
+		matrix.appendChild(app.view as unknown as Node);
+
+		// create viewport
+		const screenHeight = this.screenHeight;
+		const screenWidth = this.screenWidth;
+		const cellAspectRatio = 1 || this.microRowArray.length / this.microColArray.length;
+		const quadAspectRatio = this.screenAspectRatio * cellAspectRatio; // height / width
+		const vertexPosition =
+			quadAspectRatio > 1 // is height long?
+				? [
+						-1 / quadAspectRatio,
+						-1, // height long
+						1 / quadAspectRatio,
+						-1, // x, y
+						1 / quadAspectRatio,
+						1,
+						-1 / quadAspectRatio,
+						1
+				  ]
+				: [
+						-1,
+						-1 * quadAspectRatio, // width long
+						1,
+						-1 * quadAspectRatio, // x, y
+						1,
+						1 * quadAspectRatio,
+						-1,
+						1 * quadAspectRatio
+				  ];
+		this.worldHeight = vertexPosition[5] - vertexPosition[3];
+		this.worldWidth = vertexPosition[2] - vertexPosition[0];
+		(this as any).$viewport = new Viewport({
+			screenWidth: this.screenWidth,
+			screenHeight: this.screenHeight,
+			worldWidth: this.worldWidth,
+			worldHeight: this.worldHeight,
+			divWheel: matrix as HTMLElement,
+			interaction: app.renderer.plugins.interaction // the interaction module is important for wheel to work properly when renderer.view is placed or scaled
+		});
+
+		// add the viewport to the stage
+		app.stage.addChild((this as any).$viewport as unknown as DisplayObject);
+
+		// activate camera interaction plugins
+		(this as any).$viewport.drag().pinch().wheel();
+		(this as any).$viewport.pause = true;
+
+		// clamp camera
+		(this as any).$viewport.clampZoom({
+			maxWidth: this.worldWidth * 2,
+			maxHeight: this.worldHeight * 2
+		});
+
+		// build quad
+		const geometry = new Geometry()
+			.addAttribute('aVertexPosition', vertexPosition, 2) // the size of the attribute
+			.addAttribute(
+				'aUvs',
+				[
+					0,
+					0, // u, v
+					1,
+					0,
+					1,
+					1,
+					0,
+					1
+				], // u, v
+				2
+			) // the size of the attribute
+			.addIndex([0, 1, 2, 0, 2, 3]);
+
+		// build color array
+		const color = this.createColorArray();
+
+		// build uniforms
+		this.uniforms = {
+			// screen data
+			uScreenWidth: screenWidth,
+			uScreenHeight: screenHeight,
+
+			// geometry/viewport data
+			uWorldWidth: (this as any).$viewport.worldWidth,
+			uWorldHeight: (this as any).$viewport.worldHeight,
+			uViewportWorldWidth: (this as any).$viewport.worldWidth,
+			uViewportWorldHeight: (this as any).$viewport.worldHeight,
+
+			// grid settings
+			uGridDisplayBorder: false,
+			uGridRowDisplayLim: 60,
+			uGridRowDisplayTransition: 12,
+			uGridColDisplayLim: 60,
+			uGridColDisplayTransition: 12,
+
+			// row/col data
+			uNumRow: this.numRows,
+			uNumCol: this.numCols,
+			uMicroRow: this.uniforms.uMicroRow,
+			uMicroCol: this.uniforms.uMicroCol,
+
+			// cell element color data
+			uColor: Texture.fromBuffer(color, this.numCols, this.numRows),
+
+			uTime: 0
+		};
```

Comment:
Your uniforms don't seem to match the shader, e.g. what is `uTime`, `uGridRowDisplayLim` ...etc used for?

Reply:
some of the uniforms are only applicable for the grid, like in the case of `uGridRowDisplayLim`.

`uTime` isnt being used i should remove it.